File: KMP.java
package Algorithm;

/**
 * @Author 写你的名字
 * @Date 2022/6/11 17:56 （可以根据需要修改）
 * @Version 1.0 （版本号）
 */
public class KMP {
    public static int getIndexOf (String s1, String s2) {
        if (s1 == null || s2 == null || s2.length() < 1 || s2.length() > s1.length()) {
            return -1;
        }
        char[] str1 = s1.toCharArray();
        char[] str2 = s2.toCharArray();
        //x和y是比对的位置
        int x = 0;
        int y = 0;
        int[] next = getNextArray(str2);

        // o(N)
        while (x < str1.length && y < str2.length) {
            if (str1[x] == str2[y]) {
                x++;
                y++;
            } else if (next[y] == -1) { // y == 0
                x++;
            } else {
                y = next[y];
            }
        }
        return y == str2.length ? x - y : -1;
    }

    // 使用动态规划来求
    private static int[] getNextArray(char[] str2) {
        if (str2.length == 1) {
            return new int[] {-1};
        }
        int[] next = new int[str2.length];
        next[0] = -1;
        next[1] = 0;
        int i = 2;
        int cn = 0;
        while (i < next.length) {
            if (str2[i - 1] == str2[cn]) {
                next[i++] = ++cn;
            } else if (cn > 0) {
                cn = next[cn];
            } else {
                next[i++] = 0;
            }
        }
        return next;
    }

    public static void main(String[] args) {
        String s1 = "abcdefg";
        String s2 = "f";
        int index = getIndexOf(s1, s2);
        System.out.println(index);
    }
}

File: 算法期末考试/kmp.cpp
#include <iostream>
#include <algorithm>
#include <string>
#include <vector>

using namespace std;

// KMP算法的核心就是next数组, 不必像暴力搜索一样，目标串每次只能往前移动一个字符，而是可以根据next的指向，移动到下一个有可能完成匹配的下标处。
void get_next(string t, vector<int> &next)
{
	int i = 0, j = -1;
	next[0] = -1; // -1的含义是没有办法继续往后匹配了, 只能换一个起始点
	while (i < t.size() - 1)
	{
		if (j == -1 || t[i] == t[j])
		{
			next[++i] = ++j;
		}
		else
		{
			j = next[j];
		}
	}
}

int KMP_search(string s, string t)
{
	int i = 0, j = 0;
	vector<int> next(t.size());
	get_next(t, next);
	while (j == -1 || (i < s.size() && j < t.size()))
	{
		if (j == -1 || s[i] == t[j])
		{
			i++;
			j++;
		}
		else
		{
			j = next[j];
		}
	}
	if (j == t.size())
	{
		return i - j;
	}
	else
	{
		return -1;
	}
}

int main()
{
	string s = "aaab";
	string t = "ab";
	cout << KMP_search(s, t) << endl;
	return 0;
}
